### PR TITLE
Provider skeleton

### PR DESCRIPTION
**Pull request by leewadhams**
This is the first pass at creating the skeleton of the Provider site. It's by no means complete yet but would be good to get some eyes on it and see if there's anything that needs to be done or do differently to what we've done before.

* Copied the Employer site as a base for the structure.
* Removed the controllers/orchestrators/viewmodels for use case functionality.
* Left some of the cross cutting stuff that I thought we'd almost certainly need.
* Implemented WsFederation auth. **Still needs testing as waiting for the relying party to be set up.**

Next steps:
* Remove the shared stuff such as tag helpers, extension methods to a shared class library.
* Fix TODOs

### CODE REVIEW

File: src/Provider/Provider.Web/Configuration/EmployerRecruitSystemConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace Esfa.Recruit.Provider.Web.Configuration
+{
+    public class ProviderRecruitSystemConfiguration
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/EmployerRecruitSystemConfiguration.cs]: Maybe rename the file `ProviderRecruitSystemConfiguration.cs`

File: src/Provider/Provider.Web/Configuration/IoC.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        private static void RegisterFilterDeps(IServiceCollection services)
+        {
+            services.AddScoped<PlannedOutageResultFilter>();
+        }
+
+        private static void RegisterDynamicConfigurationDeps(IServiceCollection services)
+        {
+            services.AddSingleton<ProviderRecruitSystemConfiguration>(x => 
+                                                            {
+                                                                var svc = x.GetService<IConfigurationReader>();
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/IoC.cs]: 👍 Probably more so with provider recruit, we should start putting non sensitive and not conn strings type of config in this document.

File: src/Provider/Provider.Web/Configuration/Routing/RoutePrefixPaths.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Esfa.Recruit.Provider.Web.Configuration.Routing
+{
+    public static class RoutePrefixPaths
+    {
+        public const string AccountRoutePath = "/{ukprn:minlength(8)}/recruit";
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/Routing/RoutePrefixPaths.cs]: 👍

File: src/Provider/Provider.Web/Controllers/DashboardController.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        private readonly DashboardOrchestrator _orchestrator;
+
+        public DashboardController(DashboardOrchestrator orchestrator)
+        {
+            _orchestrator = orchestrator;
+        }
+
+        [HttpGet("", Name = RouteNames.Dashboard_Index_Get)]
+        public IActionResult Dashboard([FromRoute]string ukprn)
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Controllers/DashboardController.cs]: Should we maybe call this method index? or change the default route specified in the mvc route setup `Startup.Configure.cs`?, might be wrong in the recruit site too 😞

File: src/Provider/Provider.Web/Controllers/ErrorController.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                    break;
+            }
+
+            return View(new ErrorViewModel { StatusCode = id, RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
+        }
+
+        [Route("error/handle")]
+        public IActionResult ErrorHandler()
+        {
+            // TODO: LWA - Need to store the ProviderIdentifier somewhere!!!
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Controllers/ErrorController.cs]: Interesting, should a user be able to come into the provider site without a ukprn even for error pages?

File: src/Provider/Provider.Web/Controllers/ErrorController.cs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                    _logger.LogError(flattenedExceptions, "Aggregate exception on path: {route}", routeWhereExceptionOccurred);
+
+                    exception = flattenedExceptions.InnerExceptions.FirstOrDefault();
+                }
+
+                if (exception is InvalidStateException)
+                {
+                    _logger.LogError(exception, "Exception on path: {route}", routeWhereExceptionOccurred);
+                    AddDashboardMessage(exception.Message);
+                    return RedirectToRoute(RouteNames.Dashboard_Index_Get, new { EmployerAccountId = ukprn });
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Controllers/ErrorController.cs]: `ProviderId` or `Ukprn`

File: src/Provider/Provider.Web/Startup.Configure.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                    s.Self()
+                    .CustomSources("https://dc.services.visualstudio.com")
+                )
+                .ImageSources(s => 
+                    s.Self()
+                    .CustomSources("https://maps.googleapis.com", 
+                                    "https://www.google-analytics.com", 
+                                    "data:")
+                 )
+                .ReportUris(r => r.Uris("/ContentPolicyReport/Report")));
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Startup.Configure.cs]: 😿 - I think you removed the CSP related controller in this project.

File: src/Provider/Provider.Web/appsettings.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      "System": "Information",
+      "Microsoft": "Information"
+    }
+  },
+  "Authentication": {
+    "WtRealm": "",
+    "MetaDataAddress": ""
+  },
+  "GoogleAnalytics": {
+    "TrackingManagerCode": "GTM-WH3MK6S"
```
Comment (sachpatel) [inline src/Provider/Provider.Web/appsettings.json]: 👍

File: src/Provider/Provider.Web/Configuration/AuthenticationConfiguration.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Esfa.Recruit.Provider.Web.Configuration
+{
+    public class AuthenticationConfiguration
+    {
+        public const int SessionTimeoutMinutes = 30;
+
+        public bool IsEnabledForDev { get; set; } = false;
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/AuthenticationConfiguration.cs]: Would this bool property not be false by default?

File: src/Provider/Provider.Web/Startup.ConfigureServices.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            // Routing has to come before adding Mvc
+            services.AddRouting(opt =>
+            {
+                opt.AppendTrailingSlash = true;
+            });
+
+            // TODO: Remove if we don't follow the same pattern
+            services.Configure<RazorViewEngineOptions>(o =>
+            {
+                o.ViewLocationFormats.Add("/Views/Part1/{1}/{0}" + RazorViewEngine.ViewExtension);
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Startup.ConfigureServices.cs]: Would probably remove.

File: src/Provider/Provider.Web/Configuration/Routing/RouteNames.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Esfa.Recruit.Provider.Web.Configuration.Routing
+{
+    public static class RouteNames
+    {
+        public const string Dashboard_Index_Get = "Dashboard_Index_Get";
+        public const string Dashboard_ChangePassword = "Dashboard_ChangePassword";
+        public const string Dashboard_ChangeEmail = "Dashboard_ChangeEmail";
+
+        public const string DismissOutageMessage_Post = "DismissOutageMessage_Post";
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/Routing/RouteNames.cs]: I think you might have removed the `CookieManagerController` which handles this route.

File: src/Provider/Provider.Web/Configuration/CookieNames.cs [hunk 12]
```diff
@@ -0,0 +1,10 @@
+namespace Esfa.Recruit.Provider.Web.Configuration
+{
+    public static class CookieNames
+    {
+        public const string RecruitData = "recruit-data";
```
[the remark this answers is not in the document]
Comment (leewadhams) [reply src/Provider/Provider.Web/Configuration/CookieNames.cs]: 👍

File: src/Provider/Provider.Web/Configuration/EmployerRecruitSystemConfiguration.cs (hunk 1 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Configuration/EmployerRecruitSystemConfiguration.cs]: 👍

File: src/Provider/Provider.Web/Configuration/IoC.cs (hunk 2 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Configuration/IoC.cs]: And non-environment specific values?

File: src/Provider/Provider.Web/Controllers/DashboardController.cs (hunk 4 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Controllers/DashboardController.cs]: We use attribute routing don't we??

File: src/Provider/Provider.Web/Controllers/ErrorController.cs (hunk 5 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Controllers/ErrorController.cs]: 404

File: src/Provider/Provider.Web/Controllers/ErrorController.cs (hunk 6 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Controllers/ErrorController.cs]: 👍

File: src/Provider/Provider.Web/Startup.Configure.cs (hunk 7 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Startup.Configure.cs]: Ooops!

File: src/Provider/Provider.Web/Configuration/AuthenticationConfiguration.cs (hunk 9 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Configuration/AuthenticationConfiguration.cs]: I've actually removed this switch from provider. I didn't think that anyone actually used it.

File: src/Provider/Provider.Web/Startup.ConfigureServices.cs (hunk 10 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Startup.ConfigureServices.cs]: 👍

File: src/Provider/Provider.Web/Configuration/Routing/RouteValues.cs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Esfa.Recruit.Provider.Web.Configuration.Routing
+{
+    public static class RouteValues
+    {
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/Routing/RouteValues.cs]: Test

File: src/Provider/Provider.Web/Configuration/Routing/RouteNames.cs (hunk 11 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Configuration/Routing/RouteNames.cs]: 👍

File: src/Provider/Provider.Web/Controllers/CookieManagerController.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+            if (IsValidReturnUrl(returnUrl))
+                return Redirect(returnUrl);
+
+            return RedirectToRoute(RouteNames.Dashboard_Index_Get);
+        }
+
+        private bool IsValidReturnUrl(string returnUrl)
+        {
+            return Regex.IsMatch(returnUrl, @"^/accounts/[A-Z0-9]{6}/.*");
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Controllers/CookieManagerController.cs]: This may need amending to the provider recruit site url match.

**Comment by leewadhams** [comment]
I've now moved some thing from the Employer, Provider and Qa site into the Shared.Web project.

* Extension classes
* Model Binders
* Configuration classes
* Middleware
* Tag Helpers
* Partial views

File: src/Provider/Provider.Web/Configuration/EsfaCookieOptions.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using Esfa.Recruit.Vacancies.Client.Application.Providers;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+
+namespace Esfa.Recruit.Provider.Web.Configuration
+{
+    public static class EsfaCookieOptions
```
Comment (sachpatel) [inline src/Provider/Provider.Web/Configuration/EsfaCookieOptions.cs]: Could we move this to the RCL project? 👍 on the rest of the items moved to the RCL project.

File: src/Provider/Provider.Web/Controllers/DashboardController.cs (hunk 4 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Controllers/DashboardController.cs]: Have removed the default route pattern as we use attribute routing only currently.

File: src/Provider/Provider.Web/Controllers/CookieManagerController.cs (hunk 14 above)
Comment (leewadhams) [reply src/Provider/Provider.Web/Controllers/CookieManagerController.cs]: 👍

**Pull request merged**